osquery/filesystem/filesystem.cpp

```diff
@@ -108,111 +108,135 @@ void initializeFilesystemAPILocale() {
 }
 
 Status readFile(const fs::path& path,
-                size_t size,
-                size_t block_size,
-                bool dry_run,
-                std::function<void(std::string& buffer, size_t size)> predicate,
-                bool blocking,
-                bool log) {
-  OpenReadableFile handle(path, blocking);
-
-  if (handle.fd == nullptr || !handle.fd->isValid()) {
-    return Status::failure("Cannot open file for reading: " + path.string());
-  }
-
-  off_t file_size = static_cast<off_t>(handle.fd->size());
-
-  if (size > 0 &&
-      (handle.fd->isSpecialFile() || static_cast<off_t>(size) < file_size)) {
-    file_size = static_cast<off_t>(size);
-  }
-
-  // Apply the max byte-read based on file/link target ownership.
-  auto read_max = static_cast<off_t>(FLAGS_read_max);
-  if (file_size > read_max) {
-    if (!dry_run) {
-      auto s =
-          Status::failure("Cannot read " + path.string() +
-                          " size exceeds limit: " + std::to_string(file_size) +
-                          " > " + std::to_string(read_max));
-      if (log) {
-        LOG(WARNING) << s.getMessage();
+                std::function<void(std::string_view)> predicate,
+                bool shouldLog) {
+  PlatformFile file_handle(path, PF_OPEN_EXISTING | PF_READ | PF_NONBLOCK);
+
+  if (!file_handle.isValid()) {
+    return Status::failure("Cannot open file for reading: " +
+                           file_handle.getFilePath().string());
+  }
+
+  const std::uint64_t file_size = file_handle.size();
+
+  // Fail to read if the file is bigger than the configured limit.
+  auto status = checkFileReadLimit(file_size, path, shouldLog);
+  if (!status.ok()) {
+    return status;
+  }
+
+  const bool isSpecialFile = file_handle.isSpecialFile();
+
+  /* If the file is a regular file on disk and has no data,
+     do not attempt to read */
+  if (!isSpecialFile && file_size == 0) {
+    return Status::success();
+  }
+
+  ssize_t res = 0;
+  std::size_t total_bytes = 0;
+  char buffer[kBlockSize];
+
+  do {
+    res = file_handle.read(buffer, kBlockSize);
+
+    // EOF
+    if (res == 0) {
+      break;
+    }
+
+    if (res > 0) {
+      total_bytes += res;
+      status = checkFileReadLimit(total_bytes, path, shouldLog);
+
+      if (!status.ok()) {
+        return status;
       }
-      return s;
+
+      predicate({buffer, static_cast<std::size_t>(res)});
```

Comment:
Is it worth putting a defensive try/catch here, if there's a panic in a passed `predicate`? (Or do you think that's better handled by the callers?

Reply:
I would leave it to the callers that know what the predicate actually is and which exceptions to handle, if any.